File: 2021/day03.ts
import { readFileSync } from "fs";

const lines = readFileSync("input.txt", "utf-8").trim().split("\n");
let o2 = [...lines];
let co2 = [...lines];

let gamma = "";
let epsilon = "";

function getCount(items: string[], index: number) {
  const count = [0, 0];
  for (let item of items) {
    count[+item[index]]++;
  }
  return count;
}

// part 1
for (let i = 0; i < lines[0].length; i++) {
  const count = getCount(lines, i);

  if (count[0] > count[1]) {
    gamma += "0";
    epsilon += "1";
  } else {
    gamma += "1";
    epsilon += "0";
  }
}

const part1 = parseInt(gamma, 2) * parseInt(epsilon, 2);
console.log(part1);

// part 2
for (let i = 0; i < lines[0].length; i++) {
  const o2count = getCount(o2, i);
  const co2count = getCount(co2, i);

  if (o2.length > 1) {
    if (o2count[0] > o2count[1]) o2 = o2.filter(item => +item[i] == 0);
    else o2 = o2.filter(item => +item[i] == 1);
  }

  if (co2.length > 1) {
    if (co2count[0] > co2count[1]) co2 = co2.filter(item => +item[i] == 1);
    else co2 = co2.filter(item => +item[i] == 0);
  }
}

const part2 = parseInt(o2[0], 2) * parseInt(co2[0], 2);
console.log(part2);

File: 2021/day03.py
from copy import deepcopy

d = open("input.txt").read().splitlines()
dc = deepcopy(d)

def count(arr, index):
  c0, c1 = 0, 0
  for j in range(len(arr)):
    if arr[j][i] == "1": c1 += 1
    else: c0 += 1
  return c0, c1

def remove(arr, index, val):
  if len(arr) == 1:
    return
  return [x for x in arr if x[index] == val]

part1 = ["", ""]
for i in range(len(d[0])):
  c0, c1 = count(d, i)
  if c0 > c1:
    part1[0] += "1"
    part1[1] += "0"
  else:
    part1[0] += "0"
    part1[1] += "1"

part1 = list(map(lambda x: int(x, 2), part1))
print(part1[0] * part1[1])

for i in range(len(d[0])):
  if len(d) == 1:
    break
  c0, c1 = count(d, i)
  if c0 > c1:
    d = remove(d, i, "1")
  else:
    d = remove(d, i, "0")

for i in range(len(dc[0])):
  if len(dc) == 1:
    break
  c0, c1 = count(dc, i)
  if c0 > c1:
    dc = remove(dc, i, "0")
  else:
    dc = remove(dc, i, "1")
part2 = int(d[0], 2) * int(dc[0], 2)
print(part2)